File: Source/Domain/TranspilerRule.cs
using JsonSubTypes;
using Newtonsoft.Json;

namespace LangBuilder.Source.Domain
{

    [JsonConverter(typeof(JsonSubtypes), "Type")]
    [JsonSubtypes.KnownSubType(typeof(BlockRuleViewModel), RuleType.Block)]
    [JsonSubtypes.KnownSubType(typeof(DirectTranslationRule), RuleType.DirectTranslation)]
    [JsonSubtypes.KnownSubType(typeof(ExpressionRuleViewModel), RuleType.Expression)]
    [JsonSubtypes.KnownSubType(typeof(RuleOptionSequenceRuleViewModel), RuleType.RuleOptionSequence)]
    [JsonSubtypes.KnownSubType(typeof(RuleSequenceRuleViewModel), RuleType.RuleSequence)]
    public class TranspilerRuleViewModel
    {
        public string Name { get; set; }
        public virtual RuleType Type { get; }
    }

    public abstract class TranspilerRule
    {
        public string Name { get; set; }
        public abstract string GrammarRule { get; }
        public abstract string RuleBody { get; }
    }
}

File: src/react/langbuilder/src/api/transpiler-rule.ts
import { RuleType } from "./rule-type";

export interface TranspilerRule {
  type: RuleType;
  name: string;
  isStatement: boolean;
}

export interface DirectTranslationRule extends TranspilerRule {
  type: RuleType.DirectTranslation;
  inputSymbol: string;
  outputSymbol: string;
}

export interface ExpressionRule extends TranspilerRule {
  type: RuleType.Expression;
  expression: string;
}

export const LabelRule: ExpressionRule = {
  expression: "Label",
  name: "label",
  isStatement: false,
  type: RuleType.Expression,
};

export interface RuleOptionSequenceRule extends TranspilerRule {
  type: RuleType.RuleOptionSequence;
  rules: TranspilerRule[];
}

export interface RuleSequenceRule extends TranspilerRule {
  type: RuleType.RuleSequence;
  delimiter: string;
  rules: TranspilerRule[];
}
